elyra/contents/file_parser.py

```diff
@@ -0,0 +1,279 @@
+            raise ValueError('Unsupported file type: {}'.format(filepath))
+
+    _type = None
+
+    def __init__(self):
+        # self._operation_filepath = operation_filepath
+        # self._model = None
+
+        self.parser = None
+        self.log = log.get_logger()
+
+    @property
+    @abstractmethod
+    def type(self):
+        pass
+
+    '''
+    Removing below to keep FileParser stateless
+    @property
+    def model(self):
+        if not self._model:
+            self._model = self.get_resources()
+        return self._model
+
+    @property
+    def operation_filepath(self):
+        # TODO : Ensure this is getting the abs path
+        return self._operation_filepath
+
+    @property
+    @abstractmethod
+    def search_expressions(self) -> Dict[str, List]:
+        # a dictionary of matching regex expressions to use keyed to the json stanza
+        # Parsing for ENVs e.g. search_expressions['NAME_OF_JSON_STANZA'] = ['REGEX_EXP_1', 'REGEX_EXP_2']
+        pass
+    '''
+
+    @abstractmethod
+    def file_as_list_of_code_chunks(self) -> List[str]:
+        # A list of code `chunks` to look through. The size and length of the `chunks` is determined
+        # usually by the language. For now the concrete implementations below is by line-by-line in file
+        pass
+
+    @abstractmethod
+    def get_next_code_chunk(self, filepath):
+        # Implements a generator for lines of code for the specified operation type
+        pass
+
+    @abstractmethod
+    def get_resources(self, filepath):
+        # Returns a model dictionary of all the regex matches for each key in the regex dictionary
+        pass
+
+
+class NotebookFileParser(FileParser):
+
+    _type = "notebook"
+
+    def __init__(self):
+        # super().__init__(operation_filepath)
+        # self.language = None
+        self.parser = None
+
+    def type(self):
+        return self._type
+
+    def file_as_list_of_code_chunks(self, filepath) -> List[str]:
+        code_chunks = []
+        with open(filepath) as f:
+            self.notebook = nbformat.read(f, as_version=4)
+            language = self.notebook['metadata']['kernelspec']['language']
+
+            if language == 'python':
+                self.parser = PythonScriptParser()
+            elif language == 'r':
+                self.parser = RScriptParser()
+
+            for cell in self.notebook.cells:
+                if cell.source and cell.cell_type == "code":
+                    code_chunks.append(cell.source.split('\n'))
+        return [item for subchunk in code_chunks for item in subchunk]
+
+    def get_next_code_chunk(self, filepath):
+        with open(filepath) as f:
+            self.notebook = nbformat.read(f, as_version=4)
+            language = self.notebook['metadata']['kernelspec']['language']
+
+            if language == 'python':
+                self.parser = PythonScriptParser()
+            elif language == 'r':
+                self.parser = RScriptParser()
```

Comment:
If we move this block to the constructor or `parser` property accessor, then this method only needs to focus on getting the next chunk.

Reply:
Makes sense! I had it this way at one point and forget my reasoning for changing it. Since it requires opening the file, I'll move this portion into the constructor:

```
with open(filepath) as f:
    self.notebook = nbformat.read(f, as_version=4)
    language = self.notebook['metadata']['kernelspec']['language']

    if language == 'python':
        self.parser = PythonScriptParser()
    elif language == 'r':
        self.parser = RScriptParser()
```

And just keep the below as the get_next_chunk method
```
for cell in self.notebook.cells:
    if cell.source and cell.cell_type == "code":
        yield cell.source.split('\n')
```